### DISCUSSION TITLE
NextRequest.nextUrl.hash always returns an empty string in middleware: Is this the intended behavior?

### DISCUSSION THREAD

**Discussion by Anthony-Koinyan** [vercel/next.js#49794] [App Router]
Hi everyone,

I've been playing around with the latest stable release of Next.js (the App router to be specific) and I noticed something interesting regarding the `nextUrl` property of the request object that's available in middleware. Specifically, I noticed that whenever I try to access the `nextUrl.hash` property, I always get an empty string, regardless of whether the URL has a hash or not.

I understand that browsers don't send URL hashes to the server, so my question is whether this behavior is intentional or not. If it is, then I'm wondering why the hash property is included in the NextURL class at all, since it seems to always be empty. Wouldn't it make more sense to exclude it altogether?

I'm asking because I want to make sure that I'm not misunderstanding something about how Next.js works, and if this is indeed a bug or a design issue, I'd like to know more about it.

Thank you for your help!

**Comment by icyJoseph** [comment]
> I understand that browsers don't send URL hashes to the server

And that's the end of it. The root of the confusion is in the usage of URL as a constructor.

- https://github.com/vercel/next.js/blob/afddb6ebdade616cdd7780273be4cd28d4509890/packages/next/src/server/web/next-url.ts#L216

More precisely, the original NextURL extended the URL class:

- The impl then: https://github.com/vercel/next.js/blob/a815ba9f792a71540be88f680cd6d2d6a3f87993/packages/next/server/web/next-url.ts#L22

- The impl now: https://github.com/vercel/next.js/blob/afddb6ebdade616cdd7780273be4cd28d4509890/packages/next/src/server/web/next-url.ts#L34

And the PR that changed things: https://github.com/vercel/next.js/commit/59f7676966570093f56ff35fb0d83eb97b394b7b

That PR grouped a bunch of things into the so called internals, and stopped extending the URL object. 

And that's it, the hash is there, because it is part of the URL object. You can use the URL to build a `url` on any environment, and then at the end you can get a string representation of the URL, https://nodejs.org/api/url.html#urlhash

It is not a design issue, nor a bug. 

But, really, couldn't they just skip it? Well, then you'd get errors with `new URL(req.nextUrl)`

**Comment by JorensM** [comment]
I'm on `15.1.6` and the hash still doesn't seem to be there in the `nextUrl`

**Reply by icyJoseph** [reply]
the browser doesn't send it... not even `cURL` sends it, as soon as it sees the `#` it knows to stop there, and doesn't include it in the request.

**Reply by icyJoseph** [reply]
Just for testing sake, I made my own client, that includes the `#`, but it didn't make it to middleware, it did log the `#` in the dev info though:

<img width="571" alt="Screenshot 2025-02-27 at 21 07 06" src="https://github.com/user-attachments/assets/7ee485ac-ba11-45fd-8151-fc875b0d2591" />

Still, though, a browser won't include it

<img width="740" alt="Screenshot 2025-02-27 at 21 08 46" src="https://github.com/user-attachments/assets/610f580d-6f3e-40be-accc-61074a057f94" />

I'll debug this a bit further though, for my own entertainment.
